apps/web/pages/api/insights/download.ts

```diff
@@ -0,0 +1,104 @@
+import type { NextApiRequest, NextApiResponse } from "next";
+import stream, { Readable } from "stream";
+import { promisify } from "util";
+import { z } from "zod";
+
+import dayjs from "@calcom/dayjs";
+import { getServerSession } from "@calcom/features/auth/lib/getServerSession";
+import { EventsInsights } from "@calcom/features/insights/server/events";
+import { randomString } from "@calcom/lib/random";
+
+const pipeline = promisify(stream.pipeline);
+
+const EndpoingRequestQuerySchema = z.object({
+  startDate: z.string(),
+  endDate: z.string(),
+  teamId: z.coerce.number().optional(),
+  userId: z.coerce.number().optional(),
+  memberUserId: z.coerce.number().optional(),
+  isAll: z.coerce.boolean().optional(),
+  eventTypeId: z.coerce.number().optional(),
+});
+
+export type EndpointRequestQuery = z.infer<typeof EndpoingRequestQuerySchema>;
+
+type EndpointRequest = NextApiRequest & {
+  query: EndpointRequestQuery;
+};
+
+/**
+ * Reads a CSV string and returns it as a CSV file that can be downloaded.
+ *
+ * @param req
+ * @param res
+ *
+ * @method GET
+ *
+ * @example /api/insights/download?csvAsString=1,2,3&downloadAs=data.csv
+ */
+export const downloadCSV = async (req: EndpointRequest, res: NextApiResponse): Promise<void> => {
+  let data = {} as EndpointRequestQuery;
+  try {
+    // parse the query string using zod
+    data = EndpoingRequestQuerySchema.parse(req.query as EndpointRequestQuery);
+  } catch (e) {
+    if (e instanceof z.ZodError) {
+      res.status(400).json({ message: e.errors });
+    }
+    res.status(500).json({ message: "Internal server error" });
+    return;
+  }
+  try {
+    const { startDate, endDate, teamId, userId, memberUserId, isAll } = data;
+
+    const session = await getServerSession({ req, res });
+
+    if (!session?.user?.id) {
```

Comment:
Should we also check here if `userId === session.user.id` and if the user is authorized to access the data of this teamId? Or am I missing something and this is done somewhere else?

Reply:
You are right, I forgot about these validations since I cannot use trpc procedures here :/